Upgrade SIOP dependency

jasper-d opened the pull request:
- Add new target .net6.0 and upgrade SIOP to 6.0.3 for this new target only 
  - Add overrides for `CanGetUnflushedBytes` and `UnflushedBytes` to `WrappedWriter` for .net6.0/SIOP 6 only

NickCraver commented:
@jasper-d What is the intent overall here? I'm asking because this breaks a great deal of projects, both those depending on earlier  TFMs and especially anyone dealing with binding redirects. Most libraries can rely on the minimum required version for the APIs they need, thereby offering the most flexibility and surface area for consumers. If we upgrade dependencies to _require_ later versions, that causes a lot of real practical headaches.

Fixes are good, but limiting where this can be used and making upgrades painful is less so...and this is used by some rather large downstream package volume, so it'll cause quite a few of those headaches.

jasper-d commented:
@NickCraver Primarily I would like to use `UnflushedBytes` and `ReadAtLeastAsync`, otherwise I would need to do the accounting myself. I believe upgrading SIOP alone shouldn't break anyone, given that even 6.0.3 still supports .net461?! But maybe that's just me who lives in a happy world were only .NET 6+ is a thing.

Updating TFMs was meant as housekeeping, but shouldn't be required. I can revert those if you wish. Same for the auxiliary nuget packages.

NickCraver commented:
@jasper-d Follow-up then: where do you want to use `UnflushedBytes`? I don't see it used in the library here, only in tests, which means the library dependency need not upgrade. You can always use a newer version in your downstream project as long as there are no breaking changes in that newer version, the library doesn't need to _require_ everyone to upgrade :)

jasper-d commented:
>You can always use a newer version in your downstream project as long as there are no breaking changes in that newer version, the library doesn't need to require everyone to upgrade :)

I tried that, but when referencing this library __and__ SIOP 6.0.3 explicitly from a .NET 6 application I still get a NotSupportedException when calling UnflushedBytes. I think `WrappedWriter` would need to delegate that call to `_writer` to make it work, because that's the actual `DefaultPipeWriter` which [supports it](https://github.com/dotnet/runtime/blob/v6.0.7/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.DefaultPipeWriter.cs#L46). The abstract `PipeWriter` `WrappedWriter` derives from [doesn't](https://github.com/dotnet/runtime/blob/v6.0.7/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriter.cs#L155).

But delegating won't work (or at least I wouldn't know how), because the members simply don't exist when targeting SIOP < 6.

NickCraver commented:
@jasper-d  Gotcha - so that sounds like an argument for adding (and only adding) a net6.0 build which has a later dependency and lights up those new members, that's the approach I'd take here. Hopefully Marc has time to chime in there later, just saw this and was curious :)

mgravell commented:
The addition of a net6 build and some new overrides for new features: no problem at all. Some of the other changes look suspect and would need further reading. My advice would be: limit the PR to *exactly* what you need for this feature - leave out the TFM cull etc.

mgravell commented:
Two things caught my eye in particular (in a cursory read):

1. It looks like a `#if NET461` became a `#if NET462` in code that already had a net462 target. This appears to have the effect of changing how the net462 code operates, rather than simply removing the net461 specialization
2. That csproj change with the  very unusual package ref condition: I'm going to need a comment on what that condition is trying to achieve, and why it isn't the vanilla package ref

NickCraver commented on src/Pipelines.Sockets.Unofficial/Pipelines.Sockets.Unofficial.csproj:
```diff
@@ -30,6 +30,14 @@
+            Version="5.0.1" />
+    <!-- For .NET 6+ use System.IO.Pipelines 6.0 -->
+    <PackageReference
+            Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp' AND $([MSBuild]::VersionGreaterThanOrEquals($(TargetFrameworkVersion), '6.0'))"
```
These can be simplified to `Condition="'$(TargetFramework)'=='net6.0'"` and `Condition="'$(TargetFramework)'!='net6.0'"` like above :)

jasper-d replied:
Yeah, but then it would need to be updated when adding a new(er) target in the future (any .NET > 6). People may overlook that, just like I missed the DefineConstants for .net6. 
That being said, forgetting to update the condition here would at least break the build because https://github.com/mgravell/Pipelines.Sockets.Unofficial/blob/1594beaf8b1165a61fc51197c4358d4afa107cd8/src/Pipelines.Sockets.Unofficial/SocketConnection.cs#L428-L433 wouldn't compile anymore.

@mgravell Any preference?

NickCraver replied:
FWIW, I'd stick with the convention in the file for any PR, not adding a new one. This is harder to understand, whereas the other bits are trivial to understand and very, very rarely require modification. If there's a new TFM added, this file _has_ to be touched regardless.

jasper-d replied:
I'll update the PR accordingly when I'm back from vacation.

jasper-d commented:
I've reverted all non-necessary changes, added .net6 as a new target. and upgraded SIOP only for .net6 as suggested. I'll file a separate PR with test fixes.

mgravell commented:
thanks, sorry for delay
